bucketeer/src/main/kotlin/io/bucketeer/sdk/android/internal/event/EventCreators.kt

```diff
@@ -157,80 +157,125 @@ internal fun newErrorMetricsEvent(
+        ),
+      )
+
+      is BKTException.ForbiddenException -> Pair(
+        MetricsEventType.FORBIDDEN_ERROR,
+        MetricsEventData.ForbiddenErrorMetricsEvent(
+          apiId = apiId,
+          labels = labels,
+        ),
+      )
+
+      is BKTException.IllegalArgumentException,
+      is BKTException.IllegalStateException,
+      is BKTException.InvalidHttpMethodException,
+      -> Pair(
+        MetricsEventType.INTERNAL_SDK_ERROR,
+        MetricsEventData.InternalSdkErrorMetricsEvent(
+          apiId = apiId,
+          labels = labels,
+        ),
+      )
+
+      is BKTException.InternalServerErrorException -> Pair(
+        MetricsEventType.INTERNAL_SERVER_ERROR,
+        MetricsEventData.InternalServerErrorMetricsEvent(
+          apiId = apiId,
+          labels = labels,
+        ),
+      )
 
-  return when (type) {
-    MetricsEventType.TIMEOUT_ERROR -> MetricsEventData.TimeoutErrorMetricsEvent(
-      apiId = apiId,
-      labels = labels,
-    )
-
-    MetricsEventType.NETWORK_ERROR -> MetricsEventData.NetworkErrorMetricsEvent(
-      apiId = apiId,
-      labels = labels,
-    )
-
-    MetricsEventType.INTERNAL_SDK_ERROR -> MetricsEventData.InternalSdkErrorMetricsEvent(
-      apiId = apiId,
-      labels = labels,
-    )
-
-    MetricsEventType.BAD_REQUEST_ERROR -> MetricsEventData.BadRequestErrorMetricsEvent(
-      apiId = apiId,
-      labels = labels,
-    )
-
-    MetricsEventType.UNAUTHORIZED_ERROR -> MetricsEventData.UnauthorizedErrorMetricsEvent(
-      apiId = apiId,
-      labels = labels,
-    )
-
-    MetricsEventType.FORBIDDEN_ERROR -> MetricsEventData.ForbiddenErrorMetricsEvent(
-      apiId = apiId,
-      labels = labels,
-    )
-
-    MetricsEventType.NOT_FOUND_ERROR -> MetricsEventData.NotFoundErrorMetricsEvent(
-      apiId = apiId,
-      labels = labels,
-    )
-
-    MetricsEventType.CLIENT_CLOSED_REQUEST_ERROR -> MetricsEventData.ClientClosedRequestErrorMetricsEvent(
-      apiId = apiId,
-      labels = labels,
-    )
-
-    MetricsEventType.SERVICE_UNAVAILABLE_ERROR -> MetricsEventData.ServiceUnavailableErrorMetricsEvent(
-      apiId = apiId,
-      labels = labels,
-    )
-
-    MetricsEventType.INTERNAL_SERVER_ERROR -> MetricsEventData.InternalServerErrorMetricsEvent(
-      apiId = apiId,
-      labels = labels,
-    )
-
-    else -> MetricsEventData.UnknownErrorMetricsEvent(
-      apiId = apiId,
-      labels = labels,
-    )
-  }
+      is BKTException.NetworkException -> Pair(
+        MetricsEventType.NETWORK_ERROR,
+        MetricsEventData.NetworkErrorMetricsEvent(
+          apiId = apiId,
+          labels = labels,
+        ),
+      )
+
+      is BKTException.ServiceUnavailableException -> Pair(
+        MetricsEventType.SERVICE_UNAVAILABLE_ERROR,
+        MetricsEventData.ServiceUnavailableErrorMetricsEvent(
+          apiId = apiId,
+          labels = labels,
+        ),
+      )
+
+      is BKTException.TimeoutException -> {
+        this.let {
```

Comment:
I would change method signature to

```kotlin
internal fun newEventDataMetricEvent(
  error: BKTException,
  timestamp: Long,
  featureTag: String,
  appVersion: String,
  apiId: ApiId
): EventData.MetricsEvent {
```

This way smart-cast just works out-of-the-box.

Reply:
Let me try your suggestion

Reply:
updated in https://github.com/bucketeer-io/android-client-sdk/pull/91/commits/b30b75f1cb80136aed2d4ec70258a19f3d47d9d0